File: lib/java/src/main/java/uapi/internal/binary/ClientBinaryDecode.java
package uapi.internal.binary;

import static uapi.internal.binary.DecodeBody.decodeBody;
import static uapi.internal.binary.UnpackBody.unpackBody;

import java.util.List;
import java.util.Map;
import java.util.Objects;

import uapi.ClientBinaryStrategy;

public class ClientBinaryDecode {
    static List<Object> clientBinaryDecode(List<Object> message, Map<Integer, BinaryEncoding> recentBinaryEncoders,
            ClientBinaryStrategy binaryChecksumStrategy)
            throws BinaryEncoderUnavailableError {
        final var headers = (Map<String, Object>) message.get(0);
        final var encodedMessageBody = (Map<Object, Object>) message.get(1);
        final var binaryChecksums = (List<Integer>) headers.get("bin_");
        final var binaryChecksum = binaryChecksums.get(0);

        // If there is a binary encoding included on this message, cache it
        if (headers.containsKey("enc_")) {
            final var binaryEncoding = (Map<String, Integer>) headers.get("enc_");
            final var newBinaryEncoder = new BinaryEncoding(binaryEncoding, binaryChecksum);

            recentBinaryEncoders.put(binaryChecksum, newBinaryEncoder);
        }

        binaryChecksumStrategy.update(binaryChecksum);
        final var newCurrentChecksumStrategy = binaryChecksumStrategy.getCurrentChecksums();

        recentBinaryEncoders.entrySet().removeIf(e -> !newCurrentChecksumStrategy.contains(e.getKey()));
        final var binaryEncoder = recentBinaryEncoders.get(binaryChecksum);

        final Map<Object, Object> finalEncodedMessageBody;
        if (Objects.equals(true, headers.get("_pac"))) {
            finalEncodedMessageBody = unpackBody(encodedMessageBody);
        } else {
            finalEncodedMessageBody = encodedMessageBody;
        }

        final var messageBody = decodeBody(finalEncodedMessageBody, binaryEncoder);
        return List.of(headers, messageBody);
    }
}

File: lib/ts/src/internal/binary/ClientBinaryDecode.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { decodeBody } from "../../internal/binary/DecodeBody.js";
import { unpackBody } from "../../internal/binary/UnpackBody.js";
import { convertMapsToObjects } from "./ConvertMapsToObjects.js";
import { BinaryEncodingCache } from "./BinaryEncodingCache.js";
import { ClientBinaryStrategy } from "./ClientBinaryStrategy.js";

export function clientBinaryDecode(
    message: any[],
    binaryEncodingCache: BinaryEncodingCache,
    binaryChecksumStrategy: ClientBinaryStrategy,
): any[] {
    const headers = message[0] as Map<string, any>;
    const encodedMessageBody = message[1] as Map<any, any>;
    const binaryChecksums = headers.get("@bin_") as number[];
    const binaryChecksum = binaryChecksums[0]!;

    if (headers.has("@enc_")) {
        const binaryEncoding = headers.get("@enc_") as Map<string, number>;
        binaryEncodingCache.add(binaryChecksum, binaryEncoding);
    }

    binaryChecksumStrategy.updateChecksum(binaryChecksum);
    const newCurrentChecksumStrategy = binaryChecksumStrategy.getCurrentChecksums();

    const binaryEncoder = binaryEncodingCache.get(newCurrentChecksumStrategy[0]);

    let finalEncodedMessageBody: Map<any, any>;
    if (headers.get("@pac_") === true) {
        finalEncodedMessageBody = unpackBody(encodedMessageBody);
    } else {
        finalEncodedMessageBody = encodedMessageBody;
    }

    const messageHeader = convertMapsToObjects(headers);
    const messageBody = decodeBody(finalEncodedMessageBody, binaryEncoder);
    return [messageHeader, messageBody];
}
